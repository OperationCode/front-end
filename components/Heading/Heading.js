import React from 'react';
import { string, bool, oneOf } from 'prop-types';
import classNames from 'classnames';
import kebabCase from 'lodash/kebabCase';
import ScreenReaderOnly from 'components/ScreenReaderOnly/ScreenReaderOnly';
import LinkIcon from 'static/images/icons/FontAwesome/link-solid.svg';
import styles from './Heading.module.css';

Heading.propTypes = {
  className: string,
  hasHashLink: bool,
  hasTitleUnderline: bool,
  headingLevel: oneOf([1, 2, 3, 4, 5, 6]),
  text: string.isRequired,
};

Heading.defaultProps = {
  className: undefined,
  hasHashLink: true,
  hasTitleUnderline: false,
  headingLevel: 2,
};

function Heading({ className, hasHashLink, hasTitleUnderline, headingLevel, text }) {
  const anchorId = `${kebabCase(text)}-link`;
  const HeadingElement = `h${headingLevel}`;

  return (
    <div className={styles.headingContainer}>
      <HeadingElement
        className={classNames(className, styles.Heading, {
          [styles.underline]: hasTitleUnderline,
        })}
      >
        {hasHashLink ? (
          <div className={styles.hashLinkContainer} data-testid={`Heading Content ${anchorId}`}>
            <a id={anchorId} href={`#${anchorId}`} data-testid="Hash Link">
              <ScreenReaderOnly>Scroll Link for {text}</ScreenReaderOnly>
              <LinkIcon className={styles.icon} />
            </a>

            {text}
          </div>
        ) : (
          text
        )}
      </HeadingElement>
    </div>
  );
}

export default Heading;
