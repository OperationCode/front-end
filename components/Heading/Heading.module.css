.Heading {
  display: flex;
  text-transform: uppercase;
  text-align: center;
  margin: 1rem 0;
}

.headingContainer {
  display: flex;
  justify-content: center;
}

.underline {
  border-bottom: 4px solid var(--primary);
  margin-bottom: 1rem;
}

.hashLinkContainer {
  position: relative;
}

.hashLinkContainer > a {
  position: absolute;
  left: -2rem;
}

.icon {
  width: 1rem;
  line-height: 2.25;
  height: 2rem;
  margin: -0.2rem 0.5rem;
  outline: none;
}

/* Hide on mobile always */
@media screen and (--small-viewport) {
  .hashLinkContainer a {
    opacity: 0;
    visibility: hidden;
    display: none;
  }
}

.Heading .icon {
  display: none;
  opacity: 0;
  visibility: hidden;
  transition: 0.2s linear opacity, 0.2s linear visibility;
}

@media not screen and (--small-viewport) {
  .Heading:hover .icon,
  .Heading a:focus .icon {
    display: block;
    opacity: 1;
    visibility: visible;
    transition: 0.2s linear opacity, 0.2s linear visibility;
  }
}
