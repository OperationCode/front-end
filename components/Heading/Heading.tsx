import classNames from 'classnames';
import kebabCase from 'lodash/kebabCase';
import { ScreenReaderOnly } from '@/components/ScreenReaderOnly/ScreenReaderOnly';
import LinkIcon from '@/public/static/images/icons/FontAwesome/link-solid.svg';
import styles from './Heading.module.css';

type HeadingLevelType = 1 | 2 | 3 | 4 | 5 | 6;

export interface HeadingPropsType {
  /**
   * Text to be rendered in the heading element.
   */
  text: string;
  /**
   * Applies classnames to the base `figure` element for styling.
   */
  className?: string;
  /**
   * Applies an anchor as the base element if true.
   * @default true
   */
  hasHashLink?: boolean;
  /**
   * Displays an optional line under the title.
   * @default false
   */
  hasTitleUnderline?: boolean;
  /**
   * Sets the heading level (h1, h2, etc)
   * @default 2
   */
  headingLevel?: HeadingLevelType;
}

export function Heading({
  className,
  hasHashLink = true,
  hasTitleUnderline = false,
  headingLevel = 2,
  text,
}: HeadingPropsType) {
  const anchorId = `${kebabCase(text)}-link`;
  const HeadingElement = `h${headingLevel}` as keyof JSX.IntrinsicElements;

  return (
    <div className={styles.headingContainer}>
      <HeadingElement
        className={classNames(className, styles.Heading, {
          [styles.underline]: hasTitleUnderline,
        })}
      >
        {hasHashLink ? (
          <div className={styles.hashLinkContainer} data-testid={`Heading Content ${anchorId}`}>
            <a id={anchorId} href={`#${anchorId}`} data-testid="Hash Link">
              <ScreenReaderOnly>Scroll Link for {text}</ScreenReaderOnly>
              <LinkIcon className={styles.icon} />
            </a>

            {text}
          </div>
        ) : (
          text
        )}
      </HeadingElement>
    </div>
  );
}
