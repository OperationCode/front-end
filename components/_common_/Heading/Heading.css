.Heading {
  display: flex;
  flex-direction: row;
  justify-content: center;
  margin-bottom: 1rem;
  text-align: center;
  width: 85%;
}

.headingLines::before,
.headingLines::after {
  flex-grow: 1;
  height: 3px;
  content: "\a0";
  background-color: var(--primary);
  align-self: center;
  margin: 0 10px 5px 10px;
}

.slate {
  color: var(--slate);
}

.gray {
  color: var(--gray);
}

.white {
  color: #ffffff;
}

@media screen and (max-width: 768px) {
  .Heading {
    width: 90%;
  }
}

@media screen and (max-width: 450px) {
  .Heading {
    width: 100%;
  }
}
