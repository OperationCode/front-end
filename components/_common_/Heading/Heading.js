import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './Heading.css';

Heading.propTypes = {
  className: PropTypes.string,
  id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]), // reference for scroll anchors
  children: PropTypes.node.isRequired,
  hasHeadingLines: PropTypes.bool,
  theme: PropTypes.oneOf(['gray', 'slate', 'white']),
};

Heading.defaultProps = {
  className: '',
  id: '',
  hasHeadingLines: true,
  theme: 'gray',
};

function Heading({ children, className, hasHeadingLines, id, theme }) {
  const classes = classNames(className, styles.Heading, styles[theme], {
    [`${styles.headingLines}`]: hasHeadingLines,
  });

  return (
    <h2 className={classes} id={id}>
      {children}
    </h2>
  );
}

export default Heading;
