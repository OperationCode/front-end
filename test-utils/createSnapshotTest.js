/* eslint-env jest */
import renderer from 'react-test-renderer';

/**
 * Used to create fully-rendered snapshot test.
 * Useful for reusable component smoke tests.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
};
