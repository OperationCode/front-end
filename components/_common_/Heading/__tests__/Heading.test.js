/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import createSnapshotTest from 'test-utils/createSnapshotTest';

import Heading from '../Heading';

describe('Heading', () => {
  test('should render with just required props passed', () => {
    createSnapshotTest(<Heading>Test</Heading>);
  });

  test('should render properly with some props assigned', () => {
    createSnapshotTest(
      <Heading className="test-class" id="test-heading-1" hasHeadingLines={false}>
        Test
      </Heading>,
    );
  });

  test('should render with "slate" in classNames when theme="slate"', () => {
    const HeaderInstance = shallow(<Heading theme="slate">Test</Heading>);

    expect(HeaderInstance).toHaveClassName('slate');
  });

  test('should render without "headingLines" in classNames when hasHeadingLines={false}', () => {
    const HeaderInstance = shallow(<Heading hasHeadingLines={false}>Test</Heading>);

    expect(HeaderInstance).not.toHaveClassName('headingLines');
  });
});
