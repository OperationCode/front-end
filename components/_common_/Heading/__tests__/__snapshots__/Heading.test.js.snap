// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Heading should render properly with some props assigned 1`] = `
<h2
  className="test-class Heading gray"
  id="test-heading-1"
>
  Test
</h2>
`;

exports[`Heading should render with just required props passed 1`] = `
<h2
  className="Heading gray headingLines"
  id=""
>
  Test
</h2>
`;
